In [1]:
#!pip install selenium

In [2]:
import time
import pandas as pd
from selenium import webdriver
from selenium.webdriver.common.keys import Keys

In [3]:
import selenium
from selenium import webdriver
from selenium.webdriver.common.keys import Keys
from selenium.webdriver.chrome.service import Service
from webdriver_manager.chrome import ChromeDriverManager

# driver = webdriver.Chrome()

In [4]:
from selenium.webdriver.chrome.options import Options

In [5]:
id_ = ""
pw_ = ""

In [6]:
options = Options()
options.add_argument('--start-fullscreen')

# WebDriver Manager를 통해 ChromeDriver 자동 설치 및 설정
driver = webdriver.Chrome(service=Service(ChromeDriverManager().install()), options=options)
driver.implicitly_wait(3)


## 네이버 보안으로 인해 페이지에서 로그인이 한 번 더 필요

In [7]:
from selenium.webdriver.common.by import By

# 페이지 열기
driver.get('http://kartprice.net/')

# CSS 선택자로 요소 찾기
driver.find_element(By.CSS_SELECTOR, 'body > div.pye_wrap > div.pye_header > div.pye_header_m > div > ul > li:nth-child(2) > a').click()
time.sleep(3)

# XPath로 요소 찾기
id_s = driver.find_element(By.XPATH, '//*[@id="id"]')

NoSuchWindowException: Message: no such window: target window already closed
from unknown error: web view not found
  (Session info: chrome=129.0.6668.101)
Stacktrace:
0   chromedriver                        0x0000000103054500 cxxbridge1$str$ptr + 1917112
1   chromedriver                        0x000000010304c890 cxxbridge1$str$ptr + 1885256
2   chromedriver                        0x0000000102c5c538 cxxbridge1$string$len + 89424
3   chromedriver                        0x0000000102c37980 core::str::slice_error_fail::hbaf5d05fe3921cd2 + 3776
4   chromedriver                        0x0000000102cc7234 cxxbridge1$string$len + 526924
5   chromedriver                        0x0000000102cda5c4 cxxbridge1$string$len + 605660
6   chromedriver                        0x0000000102c95374 cxxbridge1$string$len + 322444
7   chromedriver                        0x0000000102c95fc4 cxxbridge1$string$len + 325596
8   chromedriver                        0x000000010301bd2c cxxbridge1$str$ptr + 1685732
9   chromedriver                        0x0000000103020530 cxxbridge1$str$ptr + 1704168
10  chromedriver                        0x0000000103000e08 cxxbridge1$str$ptr + 1575360
11  chromedriver                        0x0000000103020e00 cxxbridge1$str$ptr + 1706424
12  chromedriver                        0x0000000102ff1f94 cxxbridge1$str$ptr + 1514316
13  chromedriver                        0x000000010303d62c cxxbridge1$str$ptr + 1823204
14  chromedriver                        0x000000010303d7ac cxxbridge1$str$ptr + 1823588
15  chromedriver                        0x000000010304c530 cxxbridge1$str$ptr + 1884392
16  libsystem_pthread.dylib             0x000000019e2af2e4 _pthread_start + 136
17  libsystem_pthread.dylib             0x000000019e2aa0fc thread_start + 8


In [19]:
# search_list = [
# '김환기', '이우환', '박서보', '정상화', '김창열', '박수근', '천경자', '이중섭', '윤형근', '이대원', 
# '유영국', '장욱진', '김종학', '백남준', '오치균', '하종현', '도상봉', '이강소', '이왈종', '정선', 
# '김홍도', '김태호', '이상범', '남관', '이응노', '전광영', '김정희', '서세옥', '이성자', '김기창']
search_list = [
'이상범', '남관', '이응노', '전광영', '김정희', '서세옥', '이성자', '김기창']

In [30]:
def search(search_word):
    #검색할 칸을 찾아
    search = driver.find_element(By.XPATH,'//*[@id="search_val"]')
    search.clear() # 같은 사이트 여러번 크롤링했을 때의 문제 해결위함
    search.send_keys(search_word)
    time.sleep(1)
    search.send_keys(Keys.ENTER) # 엔터키를 눌러 검색 진행

In [17]:
total_title = []
total_author = []
total_year = []
total_standard = []
total_genre = []
total_technique = []
total_source = []
total_date = []
total_price = []

## 크롤링 진행

In [32]:
driver.get('http://kartprice.net/')

# 아티스트 목록에 대해 검색 및 크롤링
for artist in search_list:
    search(artist)
    successful_bids = int(driver.find_element(By.XPATH, '/html/body/div[1]/div[3]/div/div/div/div[3]/dl/dd/p[2]').text[5:])
    artist_count = 0
    if successful_bids <= 10:
        total_length = 2
    else:
        total_length = successful_bids // 10 + 2

    for j in range(1, total_length):
        page_length = len(driver.find_elements(By.CSS_SELECTOR,
                                               'body > div.pye_wrap > div:nth-child(9) > div > div > table > thead > tr'))
        page_pictures = len(driver.find_elements(By.CSS_SELECTOR, '.money'))
        print('page_pictures:', page_pictures)
        
        for picture in range(2, page_pictures + 1):  ## 게시물 크롤링
            if j == 1:
                child_param = 9
                div_param = 4
            else:
                child_param = 7
                div_param = 3

            title = driver.find_element(By.XPATH,
                                        f"/html/body/div[1]/div[{div_param}]/div/div/table/thead/tr[{picture}]/td[3]/p")
            author = driver.find_element(By.XPATH,
                                         f"/html/body/div[1]/div[{div_param}]/div/div/table/thead/tr[{picture}]/td[4]/p")
            year = driver.find_element(By.XPATH,
                                       f"/html/body/div[1]/div[{div_param}]/div/div/table/thead/tr[{picture}]/td[5]/p[1]")
            standard = driver.find_element(By.XPATH,
                                           f"/html/body/div[1]/div[{div_param}]/div/div/table/thead/tr[{picture}]/td[5]/p[2]")
            genre = driver.find_element(By.XPATH,
                                        f"/html/body/div[1]/div[{div_param}]/div/div/table/thead/tr[{picture}]/td[6]/p[1]")
            technique = driver.find_element(By.XPATH,
                                            f"/html/body/div[1]/div[{div_param}]/div/div/table/thead/tr[{picture}]/td[6]/p[2]")
            source = driver.find_element(By.XPATH,
                                         f"/html/body/div[1]/div[{div_param}]/div/div/table/thead/tr[{picture}]/td[7]/p[1]")
            date = driver.find_element(By.XPATH,
                                       f"/html/body/div[1]/div[{div_param}]/div/div/table/thead/tr[{picture}]/td[7]/p[2]")
            price = driver.find_element(By.XPATH,
                                        f"/html/body/div[1]/div[{div_param}]/div/div/table/thead/tr[{picture}]/td[8]")
            
            time.sleep(1)
            total_title.append(title.text)
            total_author.append(author.text)
            total_year.append(year.text)
            total_standard.append(standard.text)
            total_genre.append(genre.text)
            total_technique.append(technique.text)
            total_source.append(source.text)
            total_date.append(date.text)
            total_price.append(price.text)

            # 마지막 페이지에서 크롤링 관련 처리
            if picture == page_pictures:
                try:
                    driver.find_element(By.CSS_SELECTOR, "#paging_t1 > button.next").click()
                    picture = 2
                    time.sleep(2)
                except:
                    driver.get('http://kartprice.net/')
                    time.sleep(2)
                    break

    print(f'{artist} 크롤링 완료, 총 {len(total_title)}건의 크롤링 완료')

page_pictures: 10
page_pictures: 10
page_pictures: 10
page_pictures: 10
page_pictures: 10
page_pictures: 10
page_pictures: 10
page_pictures: 10
page_pictures: 10
page_pictures: 10
page_pictures: 10
page_pictures: 10
page_pictures: 10
page_pictures: 10
page_pictures: 10
page_pictures: 10
page_pictures: 10
page_pictures: 10
page_pictures: 3
이상범 크롤링 완료, 총 259건의 크롤링 완료
page_pictures: 10
page_pictures: 10
page_pictures: 10
page_pictures: 10
page_pictures: 10
page_pictures: 10
page_pictures: 10
page_pictures: 10
page_pictures: 10
page_pictures: 10
page_pictures: 4
남관 크롤링 완료, 총 352건의 크롤링 완료
page_pictures: 10
page_pictures: 10
page_pictures: 10
page_pictures: 10
page_pictures: 10
page_pictures: 10
page_pictures: 10
page_pictures: 10
page_pictures: 10
page_pictures: 10
page_pictures: 10
page_pictures: 10
page_pictures: 10
page_pictures: 7
이응노 크롤링 완료, 총 475건의 크롤링 완료
page_pictures: 10
page_pictures: 10
page_pictures: 10
page_pictures: 10
page_pictures: 10
page_pictures: 10
page_pictures: 10
page_

page_pictures: 3
이상범 크롤링 완료, 총 259건의 크롤링 완료
page_pictures: 4
남관 크롤링 완료, 총 352건의 크롤링 완료
page_pictures: 7
이응노 크롤링 완료, 총 475건의 크롤링 완료
page_pictures: 1
전광영 크롤링 완료, 총 556건의 크롤링 완료
page_pictures: 2
김정희 크롤링 완료, 총 710건의 크롤링 완료
page_pictures: 4
서세옥 크롤링 완료, 총 938건의 크롤링 완료
page_pictures: 9
이성자 크롤링 완료, 총 982건의 크롤링 완료
김기창 크롤링 완료, 총 1333건의 크롤링 완료


In [33]:
data = {"title": total_title, "author" : total_author, "year" : total_year, "standard" : total_standard,
       "genre" : total_genre, "technique" : total_technique, "source" : total_source, "date" : total_date, "price" : total_price}

df = pd.DataFrame(data, columns = ["title","author", "year", "standard", "genre", "technique", "source", "date", "price"])
df

title author       year      standard genre technique      source  \
0         추경산수    이상범  1969년(갑진)      61×127cm    회화  종이에 수묵담채         칸옥션   
1         하경산수    이상범  1969년(갑진)  135.1×68.8cm    회화  종이에 수묵담채        서울옥션   
2         추경산수    이상범  1969년(갑진)      120×34cm    회화      견본채색        아이옥션   
3       괴석에 국화    이상범       1946    125×31.5cm    회화  종이에 수묵담채    에이옥션 온라인   
4        추경산수도    이상범       1965       31×64cm    회화  종이에 수묵담채      마이아트옥션   
...        ...    ...        ...           ...   ...       ...         ...   
1328      바보화조    김기창       1982     62×71.5cm    판화       석판화    에이옥션 온라인   
1329      청록산수    김기창       1982   64.5×85.5cm    판화       석판화    에이옥션 온라인   
1330  문자도(文字圖)    김기창       1982     72.5×98cm    판화       석판화    아이옥션 온라인   
1331         무    김기창       1982     15×54.9cm    회화    종이에 수묵  아트데이옥션 온라인   
1332       군마도    김기창       1982   14.2×53.3cm    회화    종이에 수묵  아트데이옥션 온라인   

            date        price  
0     2022/12/23  ￦70,000,000  
1     2022/12/20  ￦21,000,000  
2     2022/12/06   ￦2,200,000  
3     2022/11/24   ￦1,900,000  
4     2022/11/17  ￦12,000,000  
...          ...          ...  
1328  2019/11/14     ￦260,000  
1329  2019/11/14     ￦320,000  
1330  2019/11/12     ￦140,000  
1331  2019/11/06     ￦650,000  
1332  2019/11/06     ￦550,000  

[1333 rows x 9 columns]

### 엑셀로 저장하기

In [35]:
df.to_excel(f'./crawling_files/artprice_total_20241014.xlsx', index=False)

In [36]:
df1 = pd.read_excel(f'./crawling_files/artprice_total_1.xlsx')
df2 = pd.read_excel(f'./crawling_files/artprice_total_2.xlsx')
df3 = pd.read_excel(f'./crawling_files/artprice_total_3.xlsx')

FileNotFoundError: [Errno 2] No such file or directory: './crawling_files/artprice_total_1.xlsx'

In [173]:
total_df = df1.append(df2)
total_df = total_df.append(df3)
total_df = total_df.drop_duplicates()

In [175]:
total_df.to_excel(f'./crawling_files/artprice_total.xlsx', index=False)